### DISCUSSION TITLE
Call generator on model save

### DISCUSSION THREAD

**Discussion by Guite** [eclipse-langium/langium#1321] [Q&A] [answered]
Hello,

is there a LSP interface where a generator can be called automatically as soon as a source model is saved in the editor? This would perfectly complete the replacement of the old Xtext-based DSL's behavior.

Thanks for your assistance.

**Comment by msujew** [comment, marked as answer]
Hey @Guite,

you can add a listener to the `TextDocuments#onDidSave` service of the language server:

```ts
services.shared.workspace.TextDocuments.onDidSave((event) => {
  // `event.document.uri` contains the uri of the saved document
});
```

Note that in order to fire this event, you need to adjust the capabilities of the language server. You need to create an override of the `LanguageServer` service like this:

```ts
export class CustomLanguageServer extends DefaultLanguageServer {
  protected override buildInitializeResult(params: InitializeParams): InitializeResult {
     const result = super.buildInitializeResult(params);
     result.capabilities.textDocumentSync = {
       save: true,
       change: TextDocumentSyncKind.Incremental
     };
     return result;
  }
}
```

We have an open issue to make this a bit easier: https://github.com/eclipse-langium/langium/issues/1272

**Reply by Guite** [reply]
Thank you @msujew - I will give this a try.

**Reply by Guite** [reply]
Works like a charm :+1:
